### PR TITLE
Improve search UX 

### LINKED ISSUE #149: Recordings - change display and behaviour of search results
See https://xbn8fg.axshare.com/#id=hpb2y9&p=recordings_list&view=default&g=1
(a [few tips](https://docs.google.com/document/d/1VEM0A3UmL0p4MctywCkBLk4lyh4IqejfK5JSIJlrW7E/edit) on how to use the prototype)

Change search results to "cards" instead of displaying them in a table format.

This should make the search results more usable on mobile. It’s still debatable if having them in a card format is better on desktop. If we find out a table is better, there are a couple of options:
- Change search results to a table/table look-a-like on desktop
- Allow toggling between displays (by clicking on a table view/card view button) on desktop and remember this choice

Most of the card should be clickable to navigate to the single recording page. **Don't** open the recording page in a new tab.

**Tags**
Display more information about who tagged that item. If it was the AI, show a cog. If it was a person, show the user icon. Hovering the tag displays the name of the entity/person(s) that added the tag.

**Map**
Show if available. Might need to toggle between road/topographical maps depending on the device location.

### PR DESCRIPTION
**Pull request by hardiesoft**
I'm reasonably happy with this now.  We have relative date ranges, and durations have common-presets under advanced search.  I think this should be alright to merge, and if you can create a map box free tier account, I'll throw in the static location thumbnails.

Closes  #149.

### CODE REVIEW

**Comment by mjs** [comment]
There nothing like seeing something live with real data, which of course has me wondering about few  changes (sorry Sara, and sorry Jon).

Some of the choices in the Video Tags dropdown rely on a selection in the Animals field. They used to be next to each other so this made more sense but now they're far apart. Maybe just move the Animals field to the Advanced section under Video Tags?

In terms of prioritising UI elements on the first line of each recording card, the device name is the most important, followed by the group and then recording id. Right now the recording id grabs my eye when I really care about the device name the most. Can these things be moved around?

I now realise that I look for the date and time a lot but the timestamp is kinda buried (small font, at the bottom of the card). Any ideas on how to make it easier to scan for these?

Do we really need the "Search recordings" and "Recordings" headings? I think they add clutter without much value.

Can we make the default results per page 100, instead of 300?

I love the new Date range dropdown but can we have some longer Date range options? I suggest:
- Last 30 days
- Last 90 days
- All

The cutoff for the "Short" duration option doesn't make sense as audio recordings are typically 60s and the shortest video recording are 10s. How about:
- Short (< 20s)
- Medium (20s - 2 minutes)
- Long (more than 2 minutes)

When I select a custom date range but don't enter From and To dates the tesxtual description goes wonky. For example: `All devices, audio and video recordings and all animals between Invalid Date and Invalid Date".

This gets even worse when a Duration is selected. For example: `All devices, audio and video recordings and all animals between Invalid Date and Invalid Date with durations 10 – 60s with durations 10 – 60s`. Note the double up.

There's no obvious way to select recordings of any duration. Selecting "Custom duration" appears to do that but perhaps it would be more obvious if there was a "Any" option (which would be the default).

Things I love:
- the vertical query form is so much better 
- date range dropdown
- duration dropdown
- handling of narrow displays is very nice
- textual description of the query is awesome
- recording cards look nice

I haven't looked at the code yet. All feedback so far is just this PR out.

I can sort out a mapping API for you. I'm happy for that part to land in a separate PR though.

**Comment by mjs** [comment]
Thanks for continuing with this. Let me know when you're happy with it and I'll review.

**Comment by hardiesoft** [comment]
I think it should be okay now, some code is a little ugly around serialising/deserialising search state into the url, but that was already the case, and will hopefully go away in the future.

**Comment by mjs** [comment]
I *love* the new timeline UX. So much better.

There's one little bug though. When a recording is between 12 noon and 1pm, the hour shows up as `0pm`:

image

**Comment by mjs** [comment]
Other feedback...

Recordings now don't open up in separate tabs by default (which is fine and good) but it would be nice to still able to middle click to open a recoridng in a separate tab.

Once you click on a recording, there's no way to get back to the search you came from, especially if you use the next/prev buttons a few times. I know there's more work planned in this area but is there anything quick and dirty that can be done to make this easier?

Related to this, the Done button for audio recordings now no longer does anything. It used to take you back to the search results (by closing the tab).

**Comment by hardiesoft** [comment]
I've made it so there's still an href on the recording summary, so that you can open it in a new tab if you want to.

For the issue with not being able to get back to the search you came from from a recording, this doesn't seem correct?  I can get back to the search with the browser back button, but yes, if you click through multiple recordings in the recording detail screen each one will change the browser url, so you would have to click back multiple times to get back to the search results.  Isn't this the behaviour you'd want in a browser based app?  We could definitely add a "return to search results" link or similar, but that seems like a separate issue - the done button on audio recordings could also be replaced with something like this.

**Comment by mjs** [comment]
Hi @hardiesoft ,

I'm now seeing failures when I try to run the code. Search results fail to be displayed due to the following error:

```
[Vue warn]: Error in render: "TypeError: itemDate[0][0] is undefined"

found in

---> <RecordingsView> at src/views/RecordingsView.vue
       <App> at src/App.vue
         <Root> vue.esm.js:591
TypeError: "itemDate[0][0] is undefined"
    relativeDay RecordingsView.vue:257
    render RecordingsView.vue:70
    renderList VueJS
    render RecordingsView.vue:61
```

It seems like the shape of the data being passed in to some of the formatting functions doesn't match what's expected.

**Comment by hardiesoft** [comment]
I'll check it out, can you give me the url of the search that's erroring so I can replicate it?

**Comment by mjs** [comment]
The default search against the test API server is failing for me. URL is:
```
http://localhost:8080/recordings?where=%7B%22dateRange%22%3A%7B%22relativeDateRange%22%3A-30%7D,%22recordingDateTime%22%3A%7B%22%24lt%22%3A%222019-8-8%2017%3A58%3A10%22,%22%24gt%22%3A%222019-7-9%2017%3A58%3A10%22%7D,%22duration%22%3A%7B%22%24gte%22%3A%220%22%7D%7D&limit=100&offset=0&tagMode=any
```

**Comment by mjs** [comment]
You should have access to the same recordings that I do so it should happen for you as well

**Comment by hardiesoft** [comment]
Ah, I see, when I had tested it before there were no results in the day/hour chunking for *today*, and I wasn't handling that correctly.

**Comment by mjs** [comment]
> Ah, I see, when I had tested it before there were no results in the day/hour chunking for _today_, and I wasn't handling that correctly.

I figured it would be something relatively straightforward.

**Comment by mjs** [comment]
That's fixed the search results problem. I don't be able to Shift/Ctrl-click on recordings to open them in a new tab/window though. Right clicking to get the context menu and selecting "Open in new tab" does work however.

I've tried in Firefox and Chrome.

Any ideas?

**Comment by hardiesoft** [comment]
Fixed, I was preventing the default action on the click, but really I should have been checking for if the metaKey modifier is set on the mouse event, and if so, not changing the route of the current page.

**Comment by mjs** [comment]
That's helped but now Ctrl and Shift clicking opens the recording in a new tab *as well* as opening it tab with the search results (sorry!)

**Comment by hardiesoft** [comment]
All good, turns out it's `metaKey` you want on macos, and `ctrlKey` on windows/linux!

**Comment by mjs** [comment]
Thanks, working well now!  Going through the code now.

File: src/components/RecordingSummary.vue [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        <span
+          v-for="(tag, index) in item.tags"
+          :key="index"
+          :class="[
+            'tag',
+            'badge',
+            tag.class
+          ]"
+        >
+          <!-- TODO: Change icon when info is available from API about whether tag was applied by human or AI -->
```
Comment (mjs) [inline src/components/RecordingSummary.vue]: There should already be an `automatic` field on each tag to indicate whether it's AI generated or human generated

File: src/main.js [hunk 2]
```diff
@@ -2,7 +2,8 @@ import Vue from 'vue';
 import App from './App.vue';
 import BootstrapVue from 'bootstrap-vue';
 import Vue2Filters from 'vue2-filters';
-import 'bootstrap/dist/css/bootstrap.css';
+//import 'bootstrap/dist/css/bootstrap.css';
```
Comment (mjs) [inline src/main.js]: delete?

**Comment by hardiesoft** [comment]
I'll leave these for Sara, since she knows what icons she wanted to map to each tag type I think.

**Comment by mjs** [comment]
I'm going to merge this now as it's ready enough. My comments can be addressed in a followup PR.

**Pull request merged**